lib/charms/prometheus_k8s/v0/prometheus.py

```diff
@@ -684,6 +794,77 @@ def _set_unit_ip(self, event):
                 self._charm.model.get_binding(relation).network.bind_address
             )
 
+    def _label_alert_topology(self, rule):
+        """Insert juju topology labels into an alert rule.
+
+        Args:
+            rule: a dictionary representing a Prometheus alert rule.
+
+        Returns:
+            a dictionary representing Prometheus alert rule with Juju
+            topology labels.
+        """
+        metadata = self._scrape_metadata
+        labels = rule.get("labels", {})
+        labels["juju_model"] = metadata["model"]
+        labels["juju_model_uuid"] = metadata["model_uuid"]
+        labels["juju_application"] = metadata["application"]
+        rule["labels"] = labels
+        return rule
+
+    def _label_alert_expression(self, rule):
+        """Insert juju topology filters into a Prometheus alert rule.
+
+        Args:
+            rule: a dictionary representing a Prometheus alert rule.
+
+        Returns:
+            a dictionary representing a Prometheus alert rule that filters based
+            on juju topology.
+        """
+        metadata = self._scrape_metadata
+        topology = (
+            'juju_model="{}", juju_model_uuid="{}", juju_application="{}"'.format(
+                metadata["model"], metadata["model_uuid"], metadata["application"]
+            )
+        )
+        expr = rule["expr"]  # a rule has to have an "expr"
+        expr = expr.replace("%%juju_topology%%", topology)
+        rule["expr"] = expr
+        return rule
+
+    @property
+    def _labeled_alert_groups(self):
+        """Load alert rules from rule files.
+
+        All rules from files for a consumer charm are loaded into a single
+        group. The generated name of this group includes Juju topology
+        prefixes.
+
+        Returns:
+            a list of Prometheus alert rule groups.
+        """
+        alerts = []
+        for path in Path(ALERT_RULES_PATH).glob("*.rule"):
+            if path.is_file():
+                with path.open() as rule_file:
+                    rules = yaml.safe_load(rule_file)
+                    rule = rules[0]  # each file is list of one rule
+                    rule = self._label_alert_topology(rule)
+                    rule = self._label_alert_expression(rule)
+                    alerts.append(rule)
```

Comment:
```suggestion
                with path.open() as rule_file:
                    rules = yaml.safe_load(rule_file)
                    if len(rules):
                        rule = rules[0]  # each file is list of one rule
                        rule = self._label_alert_topology(rule)
                        rule = self._label_alert_expression(rule)
                        alerts.append(rule)
```

Reply:
Even with this change, this code becomes a little hard to read because there are so many calls out to various functions. I don't think the complexity of `_label_alert_topology` and `label_alert_expression` are sufficient enough to necessarily warrant a separate function. It might just be simpler to have these elements in this main function...

Reply:
Breaking out code into small functions with intention revealing name is a stratagem to make it self documenting. The function name states the intent of the function body.

Reply:
I think a custom exception will be required here too to prevent silent failure to safe load alert rule. I need to think if it can be the same as `InvalidAletRule` (my current inclination) or it needs a different one. Again will implement, test, document and push rather than commit code from within github.

Reply:
Added `InvalidAlertRule` exception here on failure to load rule. Will push shortly.